### PR TITLE
Create mutator class for kir and refactor passes

### PR DESCRIPTION
**Pull request by csarofeen**
Also rename kir visitor/mutator classes.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp [hunk 1]
```diff
@@ -519,38 +513,64 @@ class AllocationInserter : public kir::KirVisitor {
...
+      writeInfoToGPULower(allocation, alloc_expr);
+
+      if (alloc_expr != nullptr) {
+        if (allocation.buffer->memoryType() == MemoryType::Shared) {
+          // Shared allocations go at the begining of scope
+          TORCH_INTERNAL_ASSERT(!exprs_.empty());
+          registerInsertBefore(exprs_[0], alloc_expr, nullptr);
+        } else {
+          TORCH_INTERNAL_ASSERT(allocation.alloc_place_before != nullptr);
+          // Want to add exprs in reverse order, so we insert before the
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_allocation.cpp]: nit: the comment was truncated?

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp [hunk 2]
```diff
@@ -519,38 +513,64 @@ class AllocationInserter : public kir::KirVisitor {
...
+              : &allocation.alloc_for_loop->body();
+          registerInsertBefore(
+              allocation.alloc_place_before, alloc_expr, scope);
+        }
+      }
 
-      allocs.push_back(std::move(allocation));
+      if (init_expr != nullptr) {
+        TORCH_INTERNAL_ASSERT(allocation.init_place_before != nullptr);
+        // Want to add exprs in reverse order, so we insert before the
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_allocation.cpp]: same here

**Review by naoyam — APPROVED** [review]
LGTM. Really nice to have the common mutator interface.

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp]: Should just be removed. We used to insert first the alloc, then the insertion, but with insert before, we can do this naturally with forward placed insertions. Will remove.

**Pull request merged**